driver-core/src/main/java/com/datastax/driver/core/policies/ErrorAwarePolicy.java

```diff
@@ -0,0 +1,322 @@
+/*
+ *      Copyright (C) 2012-2015 DataStax Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.datastax.driver.core.policies;
+
+import com.codahale.metrics.Meter;
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.Host;
+import com.datastax.driver.core.HostDistance;
+import com.datastax.driver.core.LatencyTracker;
+import com.datastax.driver.core.Statement;
+import com.datastax.driver.core.exceptions.ReadFailureException;
+import com.datastax.driver.core.exceptions.ReadTimeoutException;
+import com.datastax.driver.core.exceptions.UnavailableException;
+import com.datastax.driver.core.exceptions.WriteFailureException;
+import com.datastax.driver.core.exceptions.WriteTimeoutException;
+import com.google.common.collect.AbstractIterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Chainable load balancing policy that will filter out hosts based on their errors rates.
+ * <p/>
+ * The policy bases its list of hosts on a child policy on which it will compute whether this host needs to included
+ * or excluded of the query plan.
+ * <p/>
+ * The policy uses a {@link LatencyTracker} that it will register automatically on the {@link Cluster}. This tracker
+ * will not take into account successful queries and only take note of the errors produced by queries. If the
+ * accounted number of errors for a host exceeds the threshold configured in the policy, the Host will be ignored
+ * for the query plan.
+ * <p/>
+ * The policy follows the builder pattern to be created, the {@link Builder} class can be created with
+ * {@link #builder} method.
+ */
+public class ErrorAwarePolicy implements ChainableLoadBalancingPolicy {
```

Comment:
💯 , one thing we discussed that we weren't completely sure on was whether or not to exclude Hosts that pass the error threshold from the query plan, or to put them at the end of the plan.  I think based on user feedback we may eventually want to change that behavior (move to end instead of exclude like it does now), or make it configurable (not a fan of that tbh)